shanghai/core_plugins/channel.py

```diff
@@ -0,0 +1,305 @@
+        d = len(prefixed_nick) - len(nick)
+        prefixes = prefixed_nick[:d-len(prefixed_nick)]
+        modes = []
+        for prefix in prefixes:
+            i = match.group('prefix').index(prefix)
+            modes.append(match.group('mode')[i])
+
+        # add nick to channel
+        lnick = ctx.nick_lower(nick)
+
+        if lnick not in ctx.users:
+            # TODO: CAP "multi-prefix"
+            ctx.users[lnick] = User(nick, '', '')
+
+        ctx.joins[(lchannel, lnick)] = Join(
+            ctx.channels[lchannel],
+            ctx.users[lnick],
+            {'modes': modes}
+        )
+
+
+async def on_join(ctx: NetworkContext, message: Message):
+    if not message.params:
+        ctx.logger.info(f'No channel given {message}')
+        return
+    channel = message.params[0]
+    lchannel = ctx.chan_lower(channel)
+    if ctx.nick_cmp(message.prefix.name, ctx.network.nickname):
+        # add channel if not already added
+        if lchannel not in ctx.channels:
+            ctx.channels[lchannel] = Channel(channel, [])
+
+    # add nickname to list and join channel. This case is the same for both
+    # when the bot joins and when other users join.
+    if lchannel not in ctx.channels:
+        ctx.logger.warn(f'Got message from channel we\'re not in: {message!r}')
+        return
+    lnick = ctx.nick_lower(message.prefix.name)
+    if lnick not in ctx.users:
+        ctx.users[lnick] = User(*message.prefix)
+    if (lchannel, lnick) not in ctx.joins:
+        ctx.joins[(lchannel, lnick)] = Join(
+            ctx.channels[lchannel],
+            ctx.users[lnick],
+            {'modes': []}  # extra info dict, for e.g. modes
+        )
+
+
+async def on_part(ctx: NetworkContext, message: Message):
+    lchannel = ctx.chan_lower(message.params[0])
+
+    if lchannel not in ctx.channels:
+        ctx.logger.warn(f'Got message from channel we\'re not in: {message!r}')
+        return
+
+    nick = message.prefix.name
+    ctx.remove_nick_from_channel(nick, lchannel)
+
+
+async def on_kick(ctx: NetworkContext, message: Message):
+    lchannel = ctx.chan_lower(message.params[0])
+
+    if lchannel not in ctx.channels:
+        ctx.logger.warn(f'Got message from channel we\'re not in: {message!r}')
+        return
+
+    kicked = message.params[1]
+    ctx.remove_nick_from_channel(kicked, lchannel)
+
+
+async def on_nick(ctx: NetworkContext, message: Message):
+    nick = message.prefix.name
+    new_nick = message.params[0]
+
+    lnick = ctx.nick_lower(nick)
+    lnew_nick = ctx.nick_lower(new_nick)
+
+    for lkey in list(ctx.joins):
+        if ctx.nick_cmp(lkey[1], lnick):
+            lnew_key = (lkey[0], lnew_nick)
+            ctx.joins[lnew_key] = ctx.joins[lkey]
+            del ctx.joins[lkey]
+
+    if lnick in ctx.users:
+        ctx.users[lnew_nick] = ctx.users[lnick]
+        ctx.users[lnew_nick].nickname = new_nick
+        del ctx.users[lnick]
+
+    # should we do this here?
```

Comment:
I wonder. It makes more sense to include it in the place where `Network.nickname` is being tracked. Maybe a separate "identity" core plugin, which would also take care of nick collisions on join (thus currently happens in `network.py` directly).